File: packages/web/src/lib/api/workflowPublications.ts
import { apiFetch } from "./client";
import { ensureHttpsOssUrl } from "@/lib/signedUrl";

export type PublicationScope = "public" | "mine" | "used";

export interface WorkflowPublication {
  id: string;
  sourceProjectId: string;
  title: string;
  description: string;
  category: string;
  isPublic: boolean;
  /** pending | approved | rejected */
  reviewStatus?: string;
  reviewNote?: string | null;
  videoUrl: string;
  videoAssetId: string;
  /** 封面图 URL；广场默认展示，悬浮播视频 */
  coverUrl: string;
  coverAssetId: string;
  likeCount: number;
  useCount: number;
  liked: boolean;
  authorId: string;
  authorName: string;
  authorAvatar: string;
  publishedAt: string;
  updatedAt: string;
}

export interface PublicationListResult {
  items: WorkflowPublication[];
  total: number;
  page: number;
  pageSize: number;
  categories: string[];
}

export interface PublicationPreviewAsset {
  id: string;
  legacyId?: string | null;
  category: string;
  title: string;
  fileUrl: string;
  thumbnailUrl: string;
  fileType: string;
}

export interface PublicationPreview {
  publicationId: string;
  title: string;
  sourceProjectId?: string;
  flowJson: {
    nodes?: Array<Record<string, unknown>>;
    edges?: Array<Record<string, unknown>>;
    [key: string]: unknown;
  };
  assets: PublicationPreviewAsset[];
  nodeIds?: string[];
}

export interface PublishWorkflowInput {
  projectId: string;
  title: string;
  description: string;
  category: string;
  videoAssetId: string;
  coverAssetId: string;
  isPublic: boolean;
}

function normalizePublication(raw: Record<string, unknown>): WorkflowPublication {
  return {
    id: String(raw.id ?? ""),
    sourceProjectId: String(raw.sourceProjectId ?? ""),
    title: String(raw.title ?? ""),
    description: String(raw.description ?? ""),
    category: String(raw.category ?? ""),
    isPublic: Boolean(raw.isPublic),
    reviewStatus: String(raw.reviewStatus ?? "approved"),
    reviewNote: raw.reviewNote != null ? String(raw.reviewNote) : null,
    videoUrl: ensureHttpsOssUrl(String(raw.videoUrl ?? "")),
    videoAssetId: String(raw.videoAssetId ?? ""),
    coverUrl: ensureHttpsOssUrl(String(raw.coverUrl ?? "")),
    coverAssetId: String(raw.coverAssetId ?? ""),
    likeCount: Number(raw.likeCount ?? 0),
    useCount: Number(raw.useCount ?? 0),
    liked: Boolean(raw.liked),
    authorId: String(raw.authorId ?? ""),
    authorName: String(raw.authorName ?? "创作者"),
    authorAvatar: String(raw.authorAvatar ?? ""),
    publishedAt: String(raw.publishedAt ?? ""),
    updatedAt: String(raw.updatedAt ?? ""),
  };
}

/** 作品广场排序：最新 / 最热（按点赞量） */
export type PublicationSort = "latest" | "hot";

export async function listWorkflowPublications(params: {
  scope: PublicationScope;
  category?: string;
  sort?: PublicationSort;
  page?: number;
  pageSize?: number;
}): Promise<PublicationListResult> {
  const qs = new URLSearchParams();
  qs.set("scope", params.scope);
  if (params.category && params.category !== "全部") qs.set("category", params.category);
  qs.set("sort", params.sort === "hot" ? "hot" : "latest");
  qs.set("page", String(params.page ?? 1));
  qs.set("pageSize", String(params.pageSize ?? 20));
  const raw = await apiFetch<Record<string, unknown>>(
    `/api/v1/workflow-publications?${qs.toString()}`
  );
  const items = Array.isArray(raw.items)
    ? (raw.items as Record<string, unknown>[]).map(normalizePublication)
    : [];
  return {
    items,
    total: Number(raw.total ?? 0),
    page: Number(raw.page ?? 1),
    pageSize: Number(raw.pageSize ?? 20),
    categories: Array.isArray(raw.categories)
      ? (raw.categories as unknown[]).map(String)
      : [],
  };
}

export async function getWorkflowPublication(id: string): Promise<WorkflowPublication> {
  const raw = await apiFetch<Record<string, unknown>>(`/api/v1/workflow-publications/${id}`);
  return normalizePublication(raw);
}

export async function publishWorkflow(input: PublishWorkflowInput): Promise<WorkflowPublication> {
  const raw = await apiFetch<Record<string, unknown>>("/api/v1/workflow-publications", {
    method: "POST",
    body: JSON.stringify(input),
  });
  return normalizePublication(raw);
}

/** 作者删除自己的发布条目 */
export async function deleteWorkflowPublication(
  id: string
): Promise<{ id: string; deleted: boolean }> {
  return apiFetch(`/api/v1/workflow-publications/${encodeURIComponent(id)}`, {
    method: "DELETE",
  });
}

export async function togglePublicationLike(
  id: string
): Promise<{ liked: boolean; likeCount: number }> {
  return apiFetch(`/api/v1/workflow-publications/${id}/like`, { method: "POST" });
}

export async function getPublicationPreview(id: string): Promise<PublicationPreview> {
  const raw = await apiFetch<Record<string, unknown>>(
    `/api/v1/workflow-publications/${id}/preview`
  );
  const assets = Array.isArray(raw.assets)
    ? (raw.assets as Record<string, unknown>[]).map((a) => ({
        id: String(a.id ?? ""),
        legacyId: a.legacyId != null ? String(a.legacyId) : null,
        category: String(a.category ?? "image"),
        title: String(a.title ?? ""),
        fileUrl: ensureHttpsOssUrl(String(a.fileUrl ?? "")),
        thumbnailUrl: ensureHttpsOssUrl(String(a.thumbnailUrl ?? a.fileUrl ?? "")),
        fileType: String(a.fileType ?? ""),
      }))
    : [];
  return {
    publicationId: String(raw.publicationId ?? id),
    title: String(raw.title ?? ""),
    sourceProjectId: raw.sourceProjectId != null ? String(raw.sourceProjectId) : undefined,
    flowJson: (raw.flowJson as PublicationPreview["flowJson"]) || { nodes: [], edges: [] },
    assets,
    nodeIds: Array.isArray(raw.nodeIds) ? raw.nodeIds.map(String) : [],
  };
}

export async function copyWorkflowPublication(
  id: string
): Promise<{ projectId: string; workflowId: string; title: string }> {
  return apiFetch(`/api/v1/workflow-publications/${id}/copy`, { method: "POST" });
}

File: packages/api/app/api/v1/workflow_publications.py
"""工作流发布 API：发布、列表、详情、点赞、只读预览、全量复制。"""

from __future__ import annotations

from typing import Any, Literal

from fastapi import APIRouter, Depends, Query
from pydantic import BaseModel, Field
from sqlalchemy.ext.asyncio import AsyncSession

from ...core.deps import get_current_user, get_current_user_optional
from ...models.database import get_db
from ...models.user import User
from ...services import workflow_publications as pub_svc

router = APIRouter()

ScopeKey = Literal["public", "mine", "used"]
# 广场排序：最新=发布时间/更新时间；最热=点赞量
SortKey = Literal["latest", "hot"]


class PublishIn(BaseModel):
    projectId: str = Field(..., min_length=1)
    title: str = Field(..., min_length=1, max_length=20)
    description: str = Field(..., min_length=1, max_length=100)
    category: str = Field(..., min_length=1, max_length=32)
    videoAssetId: str = Field(..., min_length=1)
    coverAssetId: str = Field(..., min_length=1)
    isPublic: bool = False


@router.post("")
async def publish_workflow(
    body: PublishIn,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> dict[str, Any]:
    """发布或更新工作流到作品广场。"""
    data = await pub_svc.upsert_publication(
        db,
        current_user,
        project_id=body.projectId,
        title=body.title,
        description=body.description,
        category=body.category,
        video_asset_id=body.videoAssetId,
        cover_asset_id=body.coverAssetId,
        is_public=body.isPublic,
    )
    await db.commit()
    return data


@router.delete("/{publication_id}")
async def delete_workflow_publication(
    publication_id: str,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> dict[str, Any]:
    """作者删除自己的发布（从「我的发布」移除，可重新发布同项目）。"""
    data = await pub_svc.delete_my_publication(db, current_user, publication_id)
    await db.commit()
    return data


@router.get("")
async def list_workflow_publications(
    scope: ScopeKey = Query(default="public"),
    category: str | None = Query(default=None),
    sort: SortKey = Query(default="latest"),
    page: int = Query(default=1, ge=1),
    pageSize: int = Query(default=20, ge=1, le=50),
    db: AsyncSession = Depends(get_db),
    current_user: User | None = Depends(get_current_user_optional),
) -> dict[str, Any]:
    """列出发布条目：public / mine / used；sort=latest|hot。"""
    return await pub_svc.list_publications(
        db,
        current_user,
        scope=scope,
        category=category,
        sort=sort,
        page=page,
        page_size=pageSize,
    )


@router.get("/{publication_id}")
async def get_workflow_publication(
    publication_id: str,
    db: AsyncSession = Depends(get_db),
    current_user: User | None = Depends(get_current_user_optional),
) -> dict[str, Any]:
    """发布详情。"""
    return await pub_svc.get_publication_detail(db, current_user, publication_id)


@router.post("/{publication_id}/like")
async def like_workflow_publication(
    publication_id: str,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> dict[str, Any]:
    """点赞切换。"""
    data = await pub_svc.toggle_publication_like(db, current_user, publication_id)
    await db.commit()
    return data


@router.get("/{publication_id}/preview")
async def preview_workflow_publication(
    publication_id: str,
    db: AsyncSession = Depends(get_db),
    current_user: User | None = Depends(get_current_user_optional),
) -> dict[str, Any]:
    """只读预览：节点布局与素材 URL。"""
    return await pub_svc.get_publication_preview(db, current_user, publication_id)


@router.post("/{publication_id}/copy")
async def copy_workflow_publication(
    publication_id: str,
    db: AsyncSession = Depends(get_db),
    current_user: User = Depends(get_current_user),
) -> dict[str, Any]:
    """将公开（或本人）工作流完整复制为新项目。"""
    data = await pub_svc.copy_publication_to_new_project(db, current_user, publication_id)
    await db.commit()
    return data
